Improved errorhandling with special characters in contentpacks

### Improved errorhandling with special characters in contentpacks

As mentioned under help on our discord. Special characters produce the following stacktrace:

```
Caused by: java.lang.RuntimeException: Could not load mod landofsignals
	at net.landofrails.landofsignals.Mod.<clinit>(Mod.java:13)
	... 38 more
Caused by: java.lang.IllegalArgumentException: MALFORMED
	at java.util.zip.ZipCoder.toString(ZipCoder.java:58)
	at java.util.zip.ZipFile.getZipEntry(ZipFile.java:567)
	at java.util.zip.ZipFile.access$900(ZipFile.java:61)
	at java.util.zip.ZipFile$ZipEntryIterator.next(ZipFile.java:525)
	at java.util.zip.ZipFile$ZipEntryIterator.next(ZipFile.java:481)
	at java.util.Iterator.forEachRemaining(Iterator.java:116)
	at java.util.Spliterators$IteratorSpliterator.forEachRemaining(Spliterators.java:1801)
	at java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:512)
	at java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:502)
	at java.util.stream.ReduceOps$ReduceOp.evaluateSequential(ReduceOps.java:708)
	at java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
	at java.util.stream.ReferencePipeline.collect(ReferencePipeline.java:499)
	at net.landofrails.landofsignals.utils.contentpacks.ContentPackHandler.loadAsset(ContentPackHandler.java:54)
	at net.landofrails.landofsignals.utils.contentpacks.ContentPackHandler.loadAssets(ContentPackHandler.java:36)
	at net.landofrails.landofsignals.utils.contentpacks.ContentPackHandler.init(ContentPackHandler.java:21)
	at net.landofrails.landofsignals.LandOfSignals.commonEvent(LandOfSignals.java:46)
	at cam72cam.mod.ModCore$Proxy.event(ModCore.java:139)
	at cam72cam.mod.ModCore$ClientProxy.event(ModCore.java:152)
	at cam72cam.mod.ModCore.register(ModCore.java:49)
	at net.landofrails.landofsignals.Mod.<clinit>(Mod.java:11)
	... 38 more
```

This should either be fixed or handled properly.

--- a/src/main/java/net/landofrails/api/contentpacks/v2/deco/ContentPackDeco.java
+++ b/src/main/java/net/landofrails/api/contentpacks/v2/deco/ContentPackDeco.java
@@ -29,6 +29,7 @@ public class ContentPackDeco {
     // Processed data
     private transient Map<String, Set<String>> objTextures;
     private transient String uniqueId;
+    private Boolean isUTF8;
 
     public String getName() {
         return name;
@@ -124,6 +125,8 @@ public class ContentPackDeco {
             joiner.add("id");
         else if (id.contains(":") || id.contains("@"))
             throw new ContentPackException(String.format("ID '%s' contains ':' or '@', which is not allowed.", id));
+        if (isUTF8 == null)
+            joiner.add("isUTF8");
         if (joiner.length() > 2) {
             invalid.accept(joiner.toString());
         } else if (!base.isEmpty()) {
@@ -175,4 +178,11 @@ public class ContentPackDeco {
         }
     }
 
+    public void setUTF8(boolean isUTF8) {
+        this.isUTF8 = isUTF8;
+    }
+
+    public boolean isUTF8() {
+        return isUTF8;
+    }
 }

--- a/src/main/java/net/landofrails/api/contentpacks/v2/sign/ContentPackSign.java
+++ b/src/main/java/net/landofrails/api/contentpacks/v2/sign/ContentPackSign.java
@@ -31,6 +31,7 @@ public class ContentPackSign {
     // Processed data
     private Map<String, Set<String>> objTextures;
     private String uniqueId;
+    private Boolean isUTF8;
 
     public String getName() {
         return name;
@@ -140,6 +141,8 @@ public class ContentPackSign {
             joiner.add("name");
         if (id == null)
             joiner.add("id");
+        if (isUTF8 == null)
+            joiner.add("isUTF8");
         if (joiner.length() > 2) {
             invalid.accept(joiner.toString());
         } else if (!base.isEmpty()) {
@@ -197,4 +200,11 @@ public class ContentPackSign {
         }
     }
 
+    public void setUTF8(boolean isUTF8) {
+        this.isUTF8 = isUTF8;
+    }
+
+    public boolean isUTF8() {
+        return isUTF8;
+    }
 }

--- a/src/main/java/net/landofrails/api/contentpacks/v2/signal/ContentPackSignal.java
+++ b/src/main/java/net/landofrails/api/contentpacks/v2/signal/ContentPackSignal.java
@@ -35,6 +35,7 @@ public class ContentPackSignal {
     // Processed data
     private Map<String, Set<String>> objTextures;
     private String uniqueId;
+    private Boolean isUTF8;
 
 
     public ContentPackSignal() {
@@ -51,6 +52,7 @@ public class ContentPackSignal {
         this.itemGroupStates = itemGroupStates;
         this.references = references;
         this.metadata = metadata;
+        this.isUTF8 = true;
     }
 
     public String getName() {
@@ -159,6 +161,8 @@ public class ContentPackSignal {
             joiner.add("id");
         if (signals == null || signals.isEmpty())
             joiner.add("signals");
+        if (isUTF8 == null)
+            joiner.add("isUTF8");
         if (joiner.length() > 2) {
             invalid.accept(joiner.toString());
         } else {
@@ -261,4 +265,11 @@ public class ContentPackSignal {
         }
     }
 
+    public boolean isUTF8() {
+        return isUTF8;
+    }
+
+    public void setUTF8(boolean isUTF8) {
+        this.isUTF8 = isUTF8;
+    }
 }

--- a/src/main/java/net/landofrails/api/contentpacks/v2/signal/ContentPackSignal.java
+++ b/src/main/java/net/landofrails/api/contentpacks/v2/signal/ContentPackSignal.java
@@ -52,7 +52,6 @@ public class ContentPackSignal {
         this.itemGroupStates = itemGroupStates;
         this.references = references;
         this.metadata = metadata;
-        this.isUTF8 = true;
     }
 
     public String getName() {

--- a/src/main/java/net/landofrails/api/contentpacks/v2/signalbox/ContentPackSignalbox.java
+++ b/src/main/java/net/landofrails/api/contentpacks/v2/signalbox/ContentPackSignalbox.java
@@ -29,6 +29,7 @@ public class ContentPackSignalbox {
     // Processed data
     private transient Map<String, Set<String>> objTextures;
     private transient String uniqueId;
+    private Boolean isUTF8;
 
     public String getName() {
         return name;
@@ -124,6 +125,8 @@ public class ContentPackSignalbox {
             joiner.add("id");
         else if (id.contains(":") || id.contains("@"))
             throw new ContentPackException(String.format("ID '%s' contains ':' or '@', which is not allowed.", id));
+        if (isUTF8 == null)
+            joiner.add("isUTF8");
         if (joiner.length() > 2) {
             invalid.accept(joiner.toString());
         } else if (!base.isEmpty()) {
@@ -175,4 +178,11 @@ public class ContentPackSignalbox {
         }
     }
 
+    public void setUTF8(boolean isUTF8) {
+        this.isUTF8 = isUTF8;
+    }
+
+    public boolean isUTF8() {
+        return isUTF8;
+    }
 }

--- a/src/main/java/net/landofrails/landofsignals/LOSBlocks.java
+++ b/src/main/java/net/landofrails/landofsignals/LOSBlocks.java
@@ -380,11 +380,12 @@ public class LOSBlocks {
                 null,
                 null
         );
+        stellwandMultisignal.setUTF8(true);
         registerStellwandContent(stellwandMultisignal);
     }
 
     private static void registerSignalContentPack(ContentPackSignalPart contentPackSignalPartV1) {
-        ContentPackHandlerV1.convertToV2(contentPackSignalPartV1, false, CONTENTPACK);
+        ContentPackHandlerV1.convertToV2(contentPackSignalPartV1, false, CONTENTPACK, true);
     }
 
     private static void registerSignContentPack(String id, String name, boolean writable, Map<String, ContentPackModel[]> models) {
@@ -393,6 +394,7 @@ public class LOSBlocks {
         contentPackSign.setName(name);
         contentPackSign.setWriteable(writable);
         contentPackSign.setBase(models);
+        contentPackSign.setUTF8(true);
 
         contentPackSign.validate(missing -> {
             throw new ContentPackException(String.format(Static.MISSING_ATTRIBUTES, missing));
@@ -406,6 +408,7 @@ public class LOSBlocks {
         contentPackDeco.setId(id);
         contentPackDeco.setName(name);
         contentPackDeco.setBase(models);
+        contentPackDeco.setUTF8(true);
 
         contentPackDeco.validate(missing -> {
             throw new ContentPackException(String.format(Static.MISSING_ATTRIBUTES, missing));
@@ -420,6 +423,7 @@ public class LOSBlocks {
         contentPackDeco.setName(name);
         contentPackDeco.setBase(models);
         contentPackDeco.setRotationSteps(90f);
+        contentPackDeco.setUTF8(true);
 
         contentPackDeco.validate(missing -> {
             throw new ContentPackException(String.format(Static.MISSING_ATTRIBUTES, missing));
@@ -437,19 +441,19 @@ public class LOSBlocks {
         );
         Map<String, ContentPackSignalGroup> group = Collections.singletonMap(groupIdName, new ContentPackSignalGroup("default", states));
 
-        registerStellwandContent(
-                new ContentPackSignal(
-                        name,
-                        id,
-                        90f,
-                        LOSTabs.SIGNALS_TAB,
-                        signalModels(objPath, "general"),
-                        group,
-                        Collections.singletonMap(groupIdName, itemGroup),
-                        null,
-                        null
-                )
+        ContentPackSignal contentPackSignal = new ContentPackSignal(
+                name,
+                id,
+                90f,
+                LOSTabs.SIGNALS_TAB,
+                signalModels(objPath, "general"),
+                group,
+                Collections.singletonMap(groupIdName, itemGroup),
+                null,
+                null
         );
+        contentPackSignal.setUTF8(true);
+        registerStellwandContent(contentPackSignal);
     }
 
     private static void registerStellwandContent(ContentPackSignal contentPackSignal) {
@@ -466,6 +470,7 @@ public class LOSBlocks {
         contentPackSignalbox.setName(name);
         contentPackSignalbox.setBase(models);
         contentPackSignalbox.setRotationSteps(45f);
+        contentPackSignalbox.setUTF8(true);
 
         contentPackSignalbox.validate(missing -> {
             throw new ContentPackException(String.format(Static.MISSING_ATTRIBUTES, missing));

--- a/src/main/java/net/landofrails/landofsignals/blocks/BlockDeco.java
+++ b/src/main/java/net/landofrails/landofsignals/blocks/BlockDeco.java
@@ -60,4 +60,8 @@ public class BlockDeco extends BlockTypeEntity {
     public float getRotationSteps(String itemId) {
         return contentPackDeco.get(checkIfMissing(itemId)).getRotationSteps();
     }
+
+    public boolean isUTF8(String itemId) {
+        return contentPackDeco.get(checkIfMissing(itemId)).isUTF8();
+    }
 }

--- a/src/main/java/net/landofrails/landofsignals/blocks/BlockSignPart.java
+++ b/src/main/java/net/landofrails/landofsignals/blocks/BlockSignPart.java
@@ -62,4 +62,8 @@ public class BlockSignPart extends BlockTypeEntity {
     public float getRotationSteps(String itemId) {
         return contentPackSigns.get(checkIfMissing(itemId)).getRotationSteps();
     }
+
+    public boolean isUTF8(String itemId) {
+        return contentPackSigns.get(checkIfMissing(itemId)).isUTF8();
+    }
 }

--- a/src/main/java/net/landofrails/landofsignals/blocks/BlockSignalBox.java
+++ b/src/main/java/net/landofrails/landofsignals/blocks/BlockSignalBox.java
@@ -57,4 +57,8 @@ public class BlockSignalBox extends BlockTypeEntity {
     public void setId(String id) {
         this.id = id;
     }
+
+    public boolean isUTF8(String itemId) {
+        return contentPackSignalboxes.get(checkIfMissing(itemId)).isUTF8();
+    }
 }

--- a/src/main/java/net/landofrails/landofsignals/blocks/BlockSignalPart.java
+++ b/src/main/java/net/landofrails/landofsignals/blocks/BlockSignalPart.java
@@ -82,4 +82,8 @@ public class BlockSignalPart extends BlockTypeEntity {
     public float getRotationSteps(String id) {
         return contentPackSignals.get(id).getRotationSteps();
     }
+
+    public boolean isUTF8(String id) {
+        return contentPackSignals.get(checkIfMissing(id)).isUTF8();
+    }
 }

--- a/src/main/java/net/landofrails/landofsignals/contentpacks/ContentPackHandler.java
+++ b/src/main/java/net/landofrails/landofsignals/contentpacks/ContentPackHandler.java
@@ -1,11 +1,13 @@
 package net.landofrails.landofsignals.contentpacks;
 
-import cam72cam.mod.ModCore;
 import net.landofrails.api.contentpacks.GenericContentPack;
+import net.landofrails.landofsignals.LandOfSignals;
 import net.landofrails.stellwand.utils.exceptions.ContentPackException;
 
 import java.io.File;
 import java.io.IOException;
+import java.nio.charset.Charset;
+import java.nio.charset.StandardCharsets;
 import java.util.List;
 import java.util.Optional;
 import java.util.function.Predicate;
@@ -27,82 +29,96 @@ public class ContentPackHandler {
 
         File assetFolder = new File("./config/landofsignals");
         if (assetFolder.exists()) {
-            ModCore.Mod.info("Searching for assets..");
+            LandOfSignals.info("Searching for assets..");
 
             File[] assets = assetFolder.listFiles((dir, name) -> name.endsWith(".zip"));
 
             if (assets == null || assets.length == 0) {
-                ModCore.Mod.info("No assets found.");
+                LandOfSignals.info("No assets found.");
             } else {
                 for (File asset : assets)
-                    loadAsset(asset);
+                    loadAsset(asset, StandardCharsets.UTF_8);
             }
 
         } else {
             boolean result = assetFolder.mkdirs();
             if (result)
-                ModCore.Mod.info("Asset folder created.");
+                LandOfSignals.info("Asset folder created.");
             else
-                ModCore.Mod.warn("Couldn't create asset folder: %s", assetFolder.getPath());
+                LandOfSignals.warn("Couldn't create asset folder: %s", assetFolder.getPath());
 
         }
     }
 
-    private static void loadAsset(File asset) {
-        ModCore.Mod.info("Loading Asset: %s", asset.getAbsolutePath());
-
-        try (ZipFile zip = new ZipFile(asset)) {
+    private static void loadAsset(File asset, Charset charset) {
+        boolean isUTF8 = StandardCharsets.UTF_8.equals(charset);
+        LandOfSignals.info("Loading Asset: %s", asset.getAbsolutePath());
+        LandOfSignals.info("Encoding: %s", charset.displayName());
+        try (ZipFile zip = new ZipFile(asset, charset)) {
 
             List<ZipEntry> files = zip.stream().filter(not(ZipEntry::isDirectory)).collect(Collectors.toList());
+
             // File.getName() returns full qualified name
             Optional<ZipEntry> landofsignalsJson = files.stream().filter(f -> f.getName().endsWith("landofsignals.json"))
                     .findFirst();
             if (landofsignalsJson.isPresent()) {
-                load(zip, landofsignalsJson.get());
+                load(zip, landofsignalsJson.get(), isUTF8);
             } else {
                 throw new ContentPackException("[" + asset.getName() + "] Missing landofsignals.json");
             }
 
         } catch (IOException zipException) {
-            ModCore.Mod.error("Couldn't load asset: %s", asset.getName());
-            ModCore.Mod.error("Error: %s", zipException.getMessage());
+            LandOfSignals.error("Couldn't load asset: %s", asset.getName());
+            LandOfSignals.error("Error: %s", zipException.getMessage());
+        } catch (IllegalArgumentException illegalArgumentException) {
+            if ("MALFORMED".equals(illegalArgumentException.getMessage())) {
+                if (isUTF8) {
+                    LandOfSignals.warn("Failed loading content with UTF-8. Trying ISO_8859_1 next. Try fixing this if possible. See wiki.");
+                    loadAsset(asset, StandardCharsets.ISO_8859_1);
+                } else {
+                    LandOfSignals.error("Failed loading content with UTF-8 and ISO_8859_1. There seems to be atleast one character thats not working.");
+                    throw illegalArgumentException;
+                }
+            } else {
+                throw illegalArgumentException;
+            }
         }
     }
 
-    private static void load(ZipFile zip, ZipEntry landofsignalsJson) {
+    private static void load(ZipFile zip, ZipEntry landofsignalsJson, boolean isUTF8) {
 
         try {
             GenericContentPack genericContentPack = GenericContentPack.fromJson(
                     zip.getInputStream(zip.getEntry(landofsignalsJson.getName())));
 
             if (!genericContentPack.isValid()) {
-                ModCore.error("Failed loading ZIP named %s!\nNot all required fields have been set.", zip.getName());
+                LandOfSignals.error("Failed loading ZIP named %s!\nNot all required fields have been set.", zip.getName());
                 return;
             }
 
             String addonversion = genericContentPack.getAddonversion();
 
-            ModCore.info("Name: %s, Author: %s, Version: %s, Addonversion: %s",
+            LandOfSignals.info("Name: %s, Author: %s, Version: %s, Addonversion: %s",
                     genericContentPack.getName(),
                     genericContentPack.getAuthor(),
                     genericContentPack.getPackversion(),
                     addonversion);
 
             if ("1".equals(addonversion)) {
-                ContentPackHandlerV1.load(zip, landofsignalsJson.getName());
+                ContentPackHandlerV1.load(zip, landofsignalsJson.getName(), isUTF8);
 
             } else if ("2".equals(addonversion)) {
-                ContentPackHandlerV2.load(zip, landofsignalsJson.getName());
+                ContentPackHandlerV2.load(zip, landofsignalsJson.getName(), isUTF8);
 
             } else {
-                ModCore.error("Failed loading Contentpack named %s!\nUnsupported addonversion: %s." +
+                LandOfSignals.error("Failed loading Contentpack named %s!\nUnsupported addonversion: %s." +
                         " Either your version of LandOfSignals is not up-to-date" +
                         " or the author used an invalid addonversion.", genericContentPack.getName(), addonversion);
             }
 
 
         } catch (IOException e) {
-            ModCore.Mod.error("Error while loading Contentpack: %s", e.getMessage());
+            LandOfSignals.error("Error while loading Contentpack: %s", e.getMessage());
         }
     }
 

--- a/src/main/java/net/landofrails/landofsignals/contentpacks/ContentPackHandlerV1.java
+++ b/src/main/java/net/landofrails/landofsignals/contentpacks/ContentPackHandlerV1.java
@@ -28,7 +28,7 @@ public class ContentPackHandlerV1 {
 
     }
 
-    public static void load(ZipFile zip, String name) throws IOException {
+    public static void load(ZipFile zip, String name, boolean isUTF8) throws IOException {
         ContentPackHead contentPack = ContentPackHead.fromJson(zip.getInputStream(zip.getEntry(name)));
         // @formatter:off
         List<ZipEntry> files = zip.stream().
@@ -37,10 +37,10 @@ public class ContentPackHandlerV1 {
         // @formatter:on
 
         ModCore.info("Content for %s:", contentPack.getId());
-        addSignals(contentPack, files, zip);
+        addSignals(contentPack, files, zip, isUTF8);
     }
 
-    private static void addSignals(ContentPackHead contentPack, List<ZipEntry> files, ZipFile zip) throws IOException {
+    private static void addSignals(ContentPackHead contentPack, List<ZipEntry> files, ZipFile zip, boolean isUTF8) throws IOException {
         for (String pathToContentPackSignalSet : contentPack.getSignals()) {
             for (ZipEntry zipEntry : files) {
                 if (zipEntry.getName().equalsIgnoreCase(pathToContentPackSignalSet)) {
@@ -58,7 +58,7 @@ public class ContentPackHandlerV1 {
                                 states.add(0, null);
                                 contentPackSignalPart.setStates(states);
 
-                                convertToV2(contentPackSignalPart, true, new ContentPack(contentPack));
+                                convertToV2(contentPackSignalPart, true, new ContentPack(contentPack), isUTF8);
 
                             }
                         }
@@ -69,7 +69,7 @@ public class ContentPackHandlerV1 {
         }
     }
 
-    public static void convertToV2(ContentPackSignalPart contentPackSignalPart, boolean addToItemTranslation, ContentPack contentPack) {
+    public static void convertToV2(ContentPackSignalPart contentPackSignalPart, boolean addToItemTranslation, ContentPack contentPack, boolean isUTF8) {
 
         // ContentPackSignal
         ContentPackSignal contentPackSignal = new ContentPackSignal();
@@ -77,6 +77,7 @@ public class ContentPackHandlerV1 {
         contentPackSignal.setId(contentPackSignalPart.getId());
         contentPackSignal.setName(contentPackSignalPart.getName());
         contentPackSignal.setMetadata(Collections.singletonMap("addonversion", 1));
+        contentPackSignal.setUTF8(isUTF8);
 
         // ContentPackSignalGroup
         ContentPackSignalGroup contentPackSignalGroup = new ContentPackSignalGroup();

--- a/src/main/java/net/landofrails/landofsignals/contentpacks/ContentPackHandlerV2.java
+++ b/src/main/java/net/landofrails/landofsignals/contentpacks/ContentPackHandlerV2.java
@@ -28,7 +28,7 @@ public class ContentPackHandlerV2 {
 
     }
 
-    public static void load(ZipFile zip, String name) throws IOException {
+    public static void load(ZipFile zip, String name, boolean isUTF8) throws IOException {
         ContentPack contentPack = ContentPack.fromJson(zip.getInputStream(zip.getEntry(name)));
         ModCore.info("Content for %s:", contentPack.getId());
 
@@ -39,16 +39,16 @@ public class ContentPackHandlerV2 {
             contentPack.getContent().forEach((path, type) -> {
                 switch (type) {
                     case BLOCKSIGNAL:
-                        loadSignal(zip, path, contentPack);
+                        loadSignal(zip, path, contentPack, isUTF8);
                         break;
                     case BLOCKSIGN:
-                        loadSign(zip, path, contentPack);
+                        loadSign(zip, path, contentPack, isUTF8);
                         break;
                     case BLOCKSIGNALBOX:
-                        loadSignalbox(zip, path, contentPack);
+                        loadSignalbox(zip, path, contentPack, isUTF8);
                         break;
                     case BLOCKDECO:
-                        loadDeco(zip, path, contentPack);
+                        loadDeco(zip, path, contentPack, isUTF8);
                         break;
                     default:
                         ModCore.error("Type %s is currently not implemented in V2!", type.name());
@@ -56,14 +56,14 @@ public class ContentPackHandlerV2 {
             });
         }
         if (hasContentSets) {
-            contentPack.getContentSets().forEach(path -> loadSet(zip, path, contentPack));
+            contentPack.getContentSets().forEach(path -> loadSet(zip, path, contentPack, isUTF8));
         }
         if (!hasContent && !hasContentSets) {
             ModCore.warn("ContentPack %s does not contain any blocks");
         }
     }
 
-    private static void loadSignal(ZipFile zip, String path, ContentPack contentPack) {
+    private static void loadSignal(ZipFile zip, String path, ContentPack contentPack, boolean isUTF8) {
 
         try {
             Predicate<ZipEntry> filterRelevantZipEntry = zipEntry -> zipEntry.getName().equalsIgnoreCase(path);
@@ -71,6 +71,7 @@ public class ContentPackHandlerV2 {
 
             if (zipEntry.isPresent()) {
                 ContentPackSignal contentPackSignal = ContentPackSignal.fromJson(zip.getInputStream(zipEntry.get()));
+                contentPackSignal.setUTF8(isUTF8);
                 contentPackSignal.validate(MISSINGATTRIBUTESEXCEPTION, contentPack);
                 ModCore.info("Signal: %s", contentPackSignal.getName());
                 LOSBlocks.BLOCK_SIGNAL_PART.add(contentPackSignal);
@@ -87,13 +88,14 @@ public class ContentPackHandlerV2 {
         }
     }
 
-    private static void loadSign(ZipFile zip, String path, ContentPack contentPack) {
+    private static void loadSign(ZipFile zip, String path, ContentPack contentPack, boolean isUTF8) {
         try {
             Predicate<ZipEntry> filterRelevantZipEntry = zipEntry -> zipEntry.getName().equalsIgnoreCase(path);
             Optional<? extends ZipEntry> zipEntry = zip.stream().filter(filterRelevantZipEntry).findFirst();
 
             if (zipEntry.isPresent()) {
                 ContentPackSign contentPackSign = ContentPackSign.fromJson(zip.getInputStream(zipEntry.get()));
+                contentPackSign.setUTF8(isUTF8);
                 contentPackSign.validate(MISSINGATTRIBUTESEXCEPTION, contentPack);
                 ModCore.info("Sign: %s", contentPackSign.getName());
 
@@ -111,13 +113,14 @@ public class ContentPackHandlerV2 {
         }
     }
 
-    private static void loadSignalbox(ZipFile zip, String path, ContentPack contentPack) {
+    private static void loadSignalbox(ZipFile zip, String path, ContentPack contentPack, boolean isUTF8) {
         try {
             Predicate<ZipEntry> filterRelevantZipEntry = zipEntry -> zipEntry.getName().equalsIgnoreCase(path);
             Optional<? extends ZipEntry> zipEntry = zip.stream().filter(filterRelevantZipEntry).findFirst();
 
             if (zipEntry.isPresent()) {
                 ContentPackSignalbox contentPackSignalbox = ContentPackSignalbox.fromJson(zip.getInputStream(zipEntry.get()));
+                contentPackSignalbox.setUTF8(isUTF8);
                 contentPackSignalbox.validate(MISSINGATTRIBUTESEXCEPTION, contentPack);
                 ModCore.info("Signalbox: %s", contentPackSignalbox.getName());
 
@@ -135,13 +138,14 @@ public class ContentPackHandlerV2 {
         }
     }
 
-    private static void loadDeco(ZipFile zip, String path, ContentPack contentPack) {
+    private static void loadDeco(ZipFile zip, String path, ContentPack contentPack, boolean isUTF8) {
         try {
             Predicate<ZipEntry> filterRelevantZipEntry = zipEntry -> zipEntry.getName().equalsIgnoreCase(path);
             Optional<? extends ZipEntry> zipEntry = zip.stream().filter(filterRelevantZipEntry).findFirst();
 
             if (zipEntry.isPresent()) {
                 ContentPackDeco contentPackDeco = ContentPackDeco.fromJson(zip.getInputStream(zipEntry.get()));
+                contentPackDeco.setUTF8(isUTF8);
                 contentPackDeco.validate(MISSINGATTRIBUTESEXCEPTION, contentPack);
                 ModCore.info("Signalbox: %s", contentPackDeco.getName());
 
@@ -159,7 +163,7 @@ public class ContentPackHandlerV2 {
         }
     }
 
-    private static void loadSet(ZipFile zip, String path, ContentPack contentPack) {
+    private static void loadSet(ZipFile zip, String path, ContentPack contentPack, boolean isUTF8) {
         try {
             Predicate<ZipEntry> filterRelevantZipEntry = zipEntry -> zipEntry.getName().equalsIgnoreCase(path);
             Optional<? extends ZipEntry> zipEntry = zip.stream().filter(filterRelevantZipEntry).findFirst();
@@ -171,16 +175,16 @@ public class ContentPackHandlerV2 {
                 contentPackSet.getContent().forEach((entryPath, type) -> {
                     switch (type) {
                         case BLOCKSIGNAL:
-                            loadSignal(zip, entryPath, contentPack);
+                            loadSignal(zip, entryPath, contentPack, isUTF8);
                             break;
                         case BLOCKSIGN:
-                            loadSign(zip, entryPath, contentPack);
+                            loadSign(zip, entryPath, contentPack, isUTF8);
                             break;
                         case BLOCKSIGNALBOX:
-                            loadSignalbox(zip, entryPath, contentPack);
+                            loadSignalbox(zip, entryPath, contentPack, isUTF8);
                             break;
                         case BLOCKDECO:
-                            loadDeco(zip, entryPath, contentPack);
+                            loadDeco(zip, entryPath, contentPack, isUTF8);
                             break;
                         default:
                             ModCore.error("Type %s is currently not implemented in V2!", type.name());

--- a/src/main/java/net/landofrails/landofsignals/items/ItemDeco.java
+++ b/src/main/java/net/landofrails/landofsignals/items/ItemDeco.java
@@ -8,6 +8,7 @@ import cam72cam.mod.item.ItemStack;
 import cam72cam.mod.math.Vec3d;
 import cam72cam.mod.math.Vec3i;
 import cam72cam.mod.serialization.TagCompound;
+import cam72cam.mod.text.TextUtil;
 import cam72cam.mod.util.Facing;
 import cam72cam.mod.world.World;
 import net.landofrails.landofsignals.LOSBlocks;
@@ -24,6 +25,7 @@ import java.util.Optional;
 
 public class ItemDeco extends CustomItem {
     private static final String ITEMIDKEY = "itemId";
+    private static final String MSG_NOT_UTF8 = "message.landofsignals:non.utf.eight.items";
 
     public ItemDeco(String modID, String name) {
         super(modID, name);
@@ -91,6 +93,13 @@ public class ItemDeco extends CustomItem {
             } else {
                 tooltips.add("ID: " + itemId);
             }
+
+            boolean isUTF8 = LOSBlocks.BLOCK_DECO.isUTF8(itemId);
+            if (!isUTF8) {
+                tooltips.add("");
+                tooltips.add(TextUtil.translate(MSG_NOT_UTF8));
+            }
+
         }
         return tooltips;
     }

--- a/src/main/java/net/landofrails/landofsignals/items/ItemSignPart.java
+++ b/src/main/java/net/landofrails/landofsignals/items/ItemSignPart.java
@@ -29,6 +29,7 @@ public class ItemSignPart extends CustomItem {
     private static final String MSG_LOS_SIGNS_WRITEABLE = "message.landogsignals:signs.writeable";
     private static final String MSG_LOS_TRUE = "message.landofsignals:true";
     private static final String MSG_LOS_FALSE = "message.landofsignals:false";
+    private static final String MSG_NOT_UTF8 = "message.landofsignals:non.utf.eight.items";
 
     public ItemSignPart(String modID, String name) {
         super(modID, name);
@@ -99,6 +100,13 @@ public class ItemSignPart extends CustomItem {
             boolean writeable = LOSBlocks.BLOCK_SIGN_PART.isWritable(itemId);
             String writeableRawText = TextUtil.translate(writeable ? MSG_LOS_TRUE : MSG_LOS_FALSE);
             tooltips.add(TextUtil.translate(MSG_LOS_SIGNS_WRITEABLE, new Object[]{writeableRawText}));
+
+            boolean isUTF8 = LOSBlocks.BLOCK_SIGN_PART.isUTF8(itemId);
+            if (!isUTF8) {
+                tooltips.add("");
+                tooltips.add(TextUtil.translate(MSG_NOT_UTF8));
+            }
+
         }
         return tooltips;
     }

--- a/src/main/java/net/landofrails/landofsignals/items/ItemSignalBox.java
+++ b/src/main/java/net/landofrails/landofsignals/items/ItemSignalBox.java
@@ -8,6 +8,7 @@ import cam72cam.mod.item.ItemStack;
 import cam72cam.mod.math.Vec3d;
 import cam72cam.mod.math.Vec3i;
 import cam72cam.mod.serialization.TagCompound;
+import cam72cam.mod.text.TextUtil;
 import cam72cam.mod.util.Facing;
 import cam72cam.mod.world.World;
 import net.landofrails.landofsignals.LOSBlocks;
@@ -24,6 +25,7 @@ import java.util.Optional;
 
 public class ItemSignalBox extends CustomItem {
     private static final String ITEMIDKEY = "itemId";
+    private static final String MSG_NOT_UTF8 = "message.landofsignals:non.utf.eight.items";
 
     public ItemSignalBox(final String modID, final String name) {
         super(modID, name);
@@ -88,6 +90,13 @@ public class ItemSignalBox extends CustomItem {
             } else {
                 tooltips.add("ID: " + itemId);
             }
+
+            boolean isUTF8 = LOSBlocks.BLOCK_SIGNAL_BOX.isUTF8(itemId);
+            if (!isUTF8) {
+                tooltips.add("");
+                tooltips.add(TextUtil.translate(MSG_NOT_UTF8));
+            }
+
         }
         return tooltips;
     }

--- a/src/main/java/net/landofrails/landofsignals/items/ItemSignalPart.java
+++ b/src/main/java/net/landofrails/landofsignals/items/ItemSignalPart.java
@@ -9,6 +9,7 @@ import cam72cam.mod.math.Vec3d;
 import cam72cam.mod.math.Vec3i;
 import cam72cam.mod.serialization.TagCompound;
 import cam72cam.mod.text.PlayerMessage;
+import cam72cam.mod.text.TextUtil;
 import cam72cam.mod.util.Facing;
 import cam72cam.mod.world.World;
 import net.landofrails.landofsignals.LOSBlocks;
@@ -35,6 +36,7 @@ public class ItemSignalPart extends CustomItem {
     }
 
     private static final String ITEMIDKEY = "itemId";
+    private static final String MSG_NOT_UTF8 = "message.landofsignals:non.utf.eight.items";
 
     @Override
     public List<CreativeTab> getCreativeTabs() {
@@ -144,6 +146,13 @@ public class ItemSignalPart extends CustomItem {
             } else {
                 tooltips.add("ID: " + itemId);
             }
+
+            boolean isUTF8 = LOSBlocks.BLOCK_SIGNAL_PART.isUTF8(itemId);
+            if (!isUTF8) {
+                tooltips.add("");
+                tooltips.add(TextUtil.translate(MSG_NOT_UTF8));
+            }
+
         }
         return tooltips;
     }

--- a/src/main/resources/assets/landofsignals/lang/de_DE.lang
+++ b/src/main/resources/assets/landofsignals/lang/de_DE.lang
@@ -33,6 +33,7 @@ message.landofsignals:item.signalanimated=§4Dieses Item ist derzeit nicht verf�
 message.landogsignals:signs.writeable=Beschriftbar: %s
 message.landofsignals:true=§aJa
 message.landofsignals:false=§cNein
+message.landofsignals:non.utf.eight.items=§6WARNUNG: §eDieses Item wurde mit ISO_8859_1 dekodiert anstatt mit UTF-8 aufgrund illegaler Zeichen. Dieses Item funktioniert eventuell nicht wie angedacht.
 
 label.landofsignals:signtexteditor.savebutton=Speichern
 

--- a/src/main/resources/assets/landofsignals/lang/en_US.lang
+++ b/src/main/resources/assets/landofsignals/lang/en_US.lang
@@ -33,6 +33,7 @@ message.landofsignals:item.signalanimated=§4This item is currently not availabl
 message.landogsignals:signs.writeable=Writeable: %s
 message.landofsignals:true=§atrue
 message.landofsignals:false=§cfalse
+message.landofsignals:non.utf.eight.items=§6WARNING: §eThis Item/Block was loaded with ISO_8859_1 encoding instead of UTF-8 due to illegal characters. This item/block may not work as intended.
 
 label.landofsignals:signtexteditor.savebutton=Save
 

